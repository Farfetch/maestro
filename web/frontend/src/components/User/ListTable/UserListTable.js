import { Button, Popconfirm, Table, Tag } from "antd";
import PropTypes from "prop-types";
import { useState } from "react";

import { deleteUser } from "../../../lib/api/endpoints/user";
import { userRole as userRoleModel } from "../../../lib/api/models";
import EditUserModal from "../EditUserModal";

const userMapper = ({
  id,
  name,
  email,
  role,
  workspaceIds,
  lastLoginAt,
  createdAt,
  updatedAt
}) => ({
  key: id,
  name,
  email,
  role,
  workspaceIds,
  lastLoginAt,
  createdAt,
  updatedAt
});

const UserListTable = ({ workspaces, users, isLoading, refresh }) => {
  const [editUserId, setEditUserId] = useState(false);

  const columns = [
    {
      title: "Name",
      dataIndex: "name",
      key: "name",
      filterSearch: true,
      filters: users.map(({ id, name }) => ({
        text: name,
        value: id
      })),
      onFilter: (value, record) => record.key === value,
      render: (text, record) => (
        <>
          {text}{" "}
          {record.role === userRoleModel.ADMIN ? (
            <Tag color="blue">{userRoleModel.ADMIN}</Tag>
          ) : null}
        </>
      )
    },
    {
      title: "Email",
      dataIndex: "email",
      key: "email",
      filterSearch: true,
      filters: users.map(({ id, email }) => ({
        text: email,
        value: id
      })),
      onFilter: (value, record) => record.key === value
    },
    {
      title: "Created at",
      dataIndex: "createdAt",
      key: "createdAt",
      width: 180,
      sorter: {
        compare: (recordA, recordB) => recordA.createdAt.diff(recordB.createdAt)
      },
      defaultSortOrder: "descend",
      render: (text, record) => record.createdAt.format("L HH:mm:ss")
    },
    {
      title: "Last Login at",
      dataIndex: "lastLoginAt",
      key: "lastLoginAt",
      width: 180,
      sorter: {
        compare: (recordA, recordB) =>
          recordA.lastLoginAt
            ? recordA.lastLoginAt.diff(recordB.lastLoginAt)
            : -1
      },
      render: (text, record) =>
        record.lastLoginAt ? record.lastLoginAt.format("L HH:mm:ss") : null
    },
    {
      title: "Action",
      key: "action",
      width: 160,
      render: (text, record) => (
        <>
          <Button type="link" onClick={() => setEditUserId(record.key)}>
            Edit
          </Button>
          {!record.isDefault ? (
            <Popconfirm
              key="delete"
              placement="left"
              title={
                <>
                  <p>Are you sure you want to delete user?</p>
                  <p>The action is not reversible.</p>
                </>
              }
              okText="Yes"
              cancelText="No"
              onConfirm={async () => {
                await deleteUser(record.key);
                refresh();
              }}
            >
              <Button type="link" danger>
                Delete
              </Button>
            </Popconfirm>
          ) : null}
        </>
      )
    }
  ];

  const usersDatasource = users.map(userMapper);

  return (
    <>
      <Table
        size="small"
        isLoading={isLoading}
        dataSource={usersDatasource}
        columns={columns}
        pagination={false}
      />
      {editUserId ? (
        <EditUserModal
          isVisible={!!editUserId}
          initialvalues={users.find(({ id }) => id === editUserId)}
          workspaces={workspaces}
          users={users}
          userId={editUserId}
          onSave={() => {
            setEditUserId(false);
            refresh();
          }}
          onCancel={() => {
            setEditUserId(false);
          }}
        />
      ) : null}
    </>
  );
};

UserListTable.propTypes = {
  workspaces: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      isDefault: PropTypes.bool.isRequired,
      createdAt: PropTypes.object.isRequired,
      updatedAt: PropTypes.object.isRequired
    })
  ),
  users: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      email: PropTypes.string.isRequired,
      role: PropTypes.oneOf([userRoleModel.USER, userRoleModel.ADMIN]),
      workspaceIds: PropTypes.arrayOf(PropTypes.string),
      createdAt: PropTypes.object.isRequired,
      updatedAt: PropTypes.object.isRequired
    })
  ),
  isLoading: PropTypes.bool.isRequired
};

export default UserListTable;
