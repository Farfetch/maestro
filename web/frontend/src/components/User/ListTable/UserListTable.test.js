import { cleanup, render, waitFor } from "@testing-library/react";
import moment from "moment";
import React from "react";

import UserListTable from "./UserListTable";

afterEach(cleanup);

describe("components/User/UserListTable", () => {
  test(`should render UserListTable component`, async () => {
    const users = [
      {
        id: "user-id-1",
        name: "user 1",
        email: "user1@maestro.net",
        role: "USER",
        workspaceIds: ["1-2-3", "4-5-6"],
        createdAt: moment("2018-05-24T13:48:04.313000"),
        updatedAt: moment("2018-05-24T13:48:04.313000")
      },
      {
        id: "user-id-2",
        name: "user 2",
        email: "user2@maestro.net",
        role: "USER",
        workspaceIds: ["4-5-6"],
        createdAt: moment("2018-05-24T13:48:04.313000"),
        updatedAt: moment("2018-05-24T13:48:04.313000")
      }
    ];

    const workspaces = [
      {
        id: "1-2-3",
        name: "Workspace 1",
        createdAt: moment("2018-05-24T13:48:04.313000"),
        updatedAt: moment("2018-05-24T13:48:04.313000")
      },
      {
        id: "4-5-6",
        name: "Workspace 2",
        createdAt: moment("2018-05-24T13:48:04.313000"),
        updatedAt: moment("2018-05-24T13:48:04.313000")
      }
    ];

    const { container } = render(
      <UserListTable
        isLoading={false}
        refresh={() => null}
        users={users}
        workspaces={workspaces}
      />
    );

    await waitFor(() => {
      expect(container.outerHTML).toMatchSnapshot();
    });
  });
});
