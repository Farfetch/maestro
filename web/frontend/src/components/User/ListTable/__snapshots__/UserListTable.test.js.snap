// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/User/UserListTable should render UserListTable component 1`] = `"<div><div class=\\"ant-table-wrapper\\"><div class=\\"ant-spin-nested-loading\\"><div class=\\"ant-spin-container\\"><div class=\\"ant-table ant-table-small\\"><div class=\\"ant-table-container\\"><div class=\\"ant-table-content\\"><table style=\\"table-layout: auto;\\"><colgroup><col><col><col style=\\"width: 180px;\\"><col style=\\"width: 160px;\\"></colgroup><thead class=\\"ant-table-thead\\"><tr><th class=\\"ant-table-cell\\"><div class=\\"ant-table-filter-column\\"><span class=\\"ant-table-column-title\\">Name</span><span role=\\"button\\" tabindex=\\"-1\\" class=\\"ant-dropdown-trigger ant-table-filter-trigger\\"><span role=\\"img\\" aria-label=\\"filter\\" class=\\"anticon anticon-filter\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"filter\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M349 838c0 17.7 14.2 32 31.8 32h262.4c17.6 0 31.8-14.3 31.8-32V642H349v196zm531.1-684H143.9c-24.5 0-39.8 26.7-27.5 48l221.3 376h348.8l221.3-376c12.1-21.3-3.2-48-27.7-48z\\"></path></svg></span></span></div></th><th class=\\"ant-table-cell\\"><div class=\\"ant-table-filter-column\\"><span class=\\"ant-table-column-title\\">Email</span><span role=\\"button\\" tabindex=\\"-1\\" class=\\"ant-dropdown-trigger ant-table-filter-trigger\\"><span role=\\"img\\" aria-label=\\"filter\\" class=\\"anticon anticon-filter\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"filter\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M349 838c0 17.7 14.2 32 31.8 32h262.4c17.6 0 31.8-14.3 31.8-32V642H349v196zm531.1-684H143.9c-24.5 0-39.8 26.7-27.5 48l221.3 376h348.8l221.3-376c12.1-21.3-3.2-48-27.7-48z\\"></path></svg></span></span></div></th><th aria-sort=\\"descending\\" class=\\"ant-table-cell ant-table-column-sort ant-table-column-has-sorters\\"><div class=\\"ant-table-column-sorters\\"><span class=\\"ant-table-column-title\\">Created at</span><span class=\\"ant-table-column-sorter ant-table-column-sorter-full\\"><span class=\\"ant-table-column-sorter-inner\\"><span role=\\"img\\" aria-label=\\"caret-up\\" class=\\"anticon anticon-caret-up ant-table-column-sorter-up\\"><svg viewBox=\\"0 0 1024 1024\\" focusable=\\"false\\" data-icon=\\"caret-up\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z\\"></path></svg></span><span role=\\"img\\" aria-label=\\"caret-down\\" class=\\"anticon anticon-caret-down ant-table-column-sorter-down active\\"><svg viewBox=\\"0 0 1024 1024\\" focusable=\\"false\\" data-icon=\\"caret-down\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z\\"></path></svg></span></span></span></div></th><th class=\\"ant-table-cell\\">Action</th></tr></thead><tbody class=\\"ant-table-tbody\\"><tr data-row-key=\\"user-id-1\\" class=\\"ant-table-row ant-table-row-level-0\\"><td class=\\"ant-table-cell\\">user 1 </td><td class=\\"ant-table-cell\\">user1@maestro.net</td><td class=\\"ant-table-cell ant-table-column-sort\\">05/24/2018 13:48:04</td><td class=\\"ant-table-cell\\"><button type=\\"button\\" class=\\"ant-btn ant-btn-link\\"><span>Edit</span></button><button type=\\"button\\" class=\\"ant-btn ant-btn-link ant-btn-dangerous\\"><span>Delete</span></button></td></tr><tr data-row-key=\\"user-id-2\\" class=\\"ant-table-row ant-table-row-level-0\\"><td class=\\"ant-table-cell\\">user 2 </td><td class=\\"ant-table-cell\\">user2@maestro.net</td><td class=\\"ant-table-cell ant-table-column-sort\\">05/24/2018 13:48:04</td><td class=\\"ant-table-cell\\"><button type=\\"button\\" class=\\"ant-btn ant-btn-link\\"><span>Edit</span></button><button type=\\"button\\" class=\\"ant-btn ant-btn-link ant-btn-dangerous\\"><span>Delete</span></button></td></tr></tbody></table></div></div></div></div></div></div></div>"`;
